### PR TITLE
RCIAM-524_Extend_simplesamlphp-module-assurance

### PR DESCRIPTION
**Pull request by ioigoume**
Map OIDs to Assurance Components and Assurance Profiles

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,42 +11,67 @@ The following classes are needed to configure, in order to execute the module.
...
+                          'https://example.org/LoA#AssuranceHigh',
+                      ),
+                      'vo_test2:IdP Proxy test2' => array(
+                          'https://example.org/LoA#AssuranceLow',
+                      ),
+                  ),
+                  '/^https:\/\/example\.org\/assurance/m',                  // Pass Through values
+              ),
+              'defaultAssurance' => 'https://example.org/LowAssurance',
+              'defaultElevatedAssurance' => 'https://example.org/HighAssurance',
```
Comment (NicolasLiampotis) [inline README.md]: Assurance elevation can be configured through the `assuranceMap` option so `defaultElevetedAssurance` can be removed

File: README.md [hunk 2]
```diff
@@ -13,42 +11,67 @@ The following classes are needed to configure, in order to execute the module.
...
+              'idpTags' => array(
+                  'exampleTag01',
+                  'exampleTag02',
+              ),
+          ),
+     )
```
Comment (NicolasLiampotis) [inline README.md]: `idpTags` ==> `idpTagMap`

```
'idpTagMap' => array(
    'exampleTag01' => array(
        'https://example.org/profile/Assurance/Medium',
    ),
    'exampleTag02' => array(
        'https://example.org/profile/Assurance/High',
    ),
),
```

File: README.md [hunk 3]
```diff
@@ -13,42 +11,67 @@ The following classes are needed to configure, in order to execute the module.
...
-                'exampleTag01',
-                'exampleTag02',
-            ),
-        ),
+     authproc = array(
+          ...
+          40 => array(
+              'class' => 'assurance:DynamicAssurance',
+              'attribute' => 'eduPersonAssurance',
+              'assuranceMap' => array(
```
Comment (NicolasLiampotis) [inline README.md:52]: Rename `assuranceMap` to `attributeMap`

File: lib/Auth/Process/DynamicAssurance.php [hunk 4]
```diff
@@ -141,27 +155,46 @@ public function process(&$state)
...
         // $state['saml:sp:IdP'] will contain an entry id for the remote IdP.
         if (!empty($state['saml:sp:IdP'])) {
             $idpEntityId = $state['saml:sp:IdP'];
-            $idpMetadata = SimpleSAML_Metadata_MetaDataStorageHandler::getMetadataHandler()->getMetaData($idpEntityId, 'saml20-idp-remote');
+            $idpMetadata = SimpleSAML_Metadata_MetaDataStorageHandler::getMetadataHandler()->getMetaData(
+                $idpEntityId,
+                'saml20-idp-remote'
+            );
         } else {
             $idpEntityId = $state['Source']['entityid'];
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/DynamicAssurance.php]: Use `idpTagMap` to evaluate assurance

File: README.md [hunk 5]
```diff
@@ -1,10 +1,8 @@
 # simplesamlphp-module-assurance
 A SimpleSAMLphp module for determining and indicating the Level of Assurance (LoA) of an authentication event.
 Specifically the module determines the LoA baced on:
-* `candidates`, if the represented LoA attribute of the authenticating user contains any value of the allowed values, then the module will skip execution.
+* `assuranceMap`, if the represented LoA attributes, of the authenticating user, contain any of the list of the allowed values, then the module will append these values into the State.
```
Comment (NicolasLiampotis) [inline README.md]: `attributeMap`, a map whose keys identify attribute names whose values can be mapped to assurance values.  represented LoA attributes, of the authenticating user, contain any of the list of the allowed values, then the module will append these values into the State.
Defaults to XXX

**Review by NicolasLiampotis — CHANGES_REQUESTED** [review]

File: README.md [hunk 6]
```diff
@@ -13,42 +11,67 @@ The following classes are needed to configure, in order to execute the module.
...
+                'https://refeds.org/assurance/IAP/low',
+                'https://refeds.org/assurance/IAP/medium',
+            ),
+        ),
+        '/^https:\/\/refeds\.org\/assurance/m',                 // REFEDS passthrough values
+        '/^https:\/\/aarc-community\.org\/assurance/m',         // AARC passthrough values
+    );
+  ```
+  :warning: The configured Assurance Map array is merged recursively with the default one from above.
+  * `defaultAssurance`: _Optional_, a string to use as the default value of LoA. Defaults to `https://www.example.org/low`.
```
Comment (NicolasLiampotis) [inline README.md]: Should default to null, i.e. no assurance value is assigned by default

File: README.md (hunk 5 above)
Comment (ioigoume) [reply README.md]: Done

File: README.md (hunk 6 above)
Comment (ioigoume) [reply README.md]: Done

File: README.md (hunk 3 above)
Comment (ioigoume) [reply README.md:52]: Done

File: README.md (hunk 2 above)
Comment (ioigoume) [reply README.md]: Done

File: README.md (hunk 1 above)
Comment (ioigoume) [reply README.md]: Done

File: README.md [hunk 7]
```diff
@@ -1,10 +1,8 @@
...
-* `idpTags`, Identity Providers (IdPs) that are providing a higher level of assuranse for the authincating user can be marked, by adding the attribute `'tags' => array (0 => 'exampleTag',)` in the IdP's metadata in `saml20-idp-remote.php`.
-* `idpPolicies`, if the 'eduPersonAssurance' attribute of the authenticating user contains any value of this list, then the user will grand higher LoA.
-* `entitlements`, if the 'eduPersonEntitlement' attribute of the authenticating user contains any value of this list, then the user will grant higher LoA.
+Specifically the module determines the LoA based on:
+* `attibuteMap`, a map whose `keys` identify attribute names whose values can be mapped to assurance values. Represented LoA attributes, of the authenticating user, contain any of the list of the allowed values, then the module will append these values into the State.
+* `idpTagMap`, a map whose `keys` identify IdP Tags which can be mapped to assurance values. IdP Metadata, of the authenticating user, contain any of the list of the allowed values, then the module will append these values into the State.
```
Comment (NicolasLiampotis) [inline README.md]: This should be removed since this is only relevant for the DynamicAssurance filter.

File: README.md [hunk 8]
```diff
@@ -13,42 +11,67 @@ The following classes are needed to configure, in order to execute the module.
 ### DynamicAssurance
```
Comment (NicolasLiampotis) [inline README.md:8]: Add introduction to filter:
```
This filter evaluates the assurance values based on specific attribute mapping rules. The evaluation process can optionally take into account the tag(s) of the authenticating Identity Provider.
```

File: lib/Auth/Process/DynamicAssurance.php [hunk 9]
```diff
@@ -141,67 +152,69 @@ public function process(&$state)
...
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/DynamicAssurance.php]: Retrieve the IdP tags *only* if the idpTagMap is non-empty

File: README.md [hunk 10]
```diff
@@ -13,42 +11,67 @@ The following classes are needed to configure, in order to execute the module.
...
+                'https://refeds.org/assurance/IAP/low',
+                'https://refeds.org/assurance/IAP/medium',
+            ),
+        ),
+        '/^https:\/\/refeds\.org\/assurance/m',                 // REFEDS passthrough values
+        '/^https:\/\/aarc-community\.org\/assurance/m',         // AARC passthrough values
+    );
+  ```
+  :warning: The configured Assurance Map array is merged recursively with the default one from above.
+  * `defaultAssurance`: _Optional_, a string to use as the default value of LoA. Defaults to `null`.
```
Comment (NicolasLiampotis) [inline README.md]: Turn this into an array to support assurance frameworks such as RAF which specify multiple assurance component values. New definition:
```
`defaultAssurance`: _Optional_, an array containing assurance value(s) to set by default when no assurance information is available. Defaults to `array()`, i.e. no assurance values are added by default.
```

File: README.md [hunk 11]
```diff
@@ -13,42 +11,67 @@ The following classes are needed to configure, in order to execute the module.
...
+                      'vo_test:IdP Proxy test' => array(
+                          'https://example.org/LoA#AssuranceHigh',
+                      ),
+                      'vo_test2:IdP Proxy test2' => array(
+                          'https://example.org/LoA#AssuranceLow',
+                      ),
+                  ),
+                  '/^https:\/\/example\.org\/assurance/m',                  // Pass Through values
+              ),
+              'defaultAssurance' => 'https://example.org/LowAssurance',
```
Comment (NicolasLiampotis) [inline README.md:81]:
```
'defaultAssurance' => array(
    'https://example.org/LowAssurance',
),
```

File: lib/Auth/Process/DynamicAssurance.php [hunk 12]
```diff
@@ -1,40 +1,53 @@
...
+ *                      'vo_test:IdP Proxy test' => array(
+ *                          'https://example.org/LoA#AssuranceHigh',
+ *                      ),
+ *                      'vo_test2:IdP Proxy test2' => array(
+ *                          'https://example.org/LoA#AssuranceLow',
+ *                      ),
+ *                  ),
+ *                  '/^https:\/\/example\.org\/assurance/m',                    // Pass Through values
  *              ),
  *              'defaultAssurance' => 'https://example.org/LowAssurance',
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/DynamicAssurance.php]:
```
'defaultAssurance' => array(
    'https://example.org/LowAssurance',
),
```

File: lib/Auth/Process/DynamicAssurance.php [hunk 13]
```diff
@@ -44,62 +57,54 @@ class sspmod_assurance_Auth_Process_DynamicAssurance extends SimpleSAML_Auth_Pro
...
     /**
      * @var string
      */
-    private $defaultAssurance = 'https://www.example.org/low';
-
-    /**
-     * @var string
-     */
-    private $defaultElevatedAssurance = 'https://www.example.org/Substantial';
+    private $defaultAssurance = null;
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/DynamicAssurance.php]:
```
private $defaultAssurance = array();
```

File: lib/Auth/Process/DynamicAssurance.php [hunk 14]
```diff
@@ -44,62 +57,54 @@ class sspmod_assurance_Auth_Process_DynamicAssurance extends SimpleSAML_Auth_Pro
...
-     */
-    private $idpTags = array();
+    private $idpTagMap = array();
 
     /**
      * @var string[]
      */
     private $config_param_str = array(
         'attribute',
         'defaultAssurance',
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/DynamicAssurance.php]: Remove defaultAssurance

File: lib/Auth/Process/DynamicAssurance.php [hunk 15]
```diff
@@ -44,62 +57,54 @@ class sspmod_assurance_Auth_Process_DynamicAssurance extends SimpleSAML_Auth_Pro
...
-        'defaultElevatedAssurance',
     );
 
     /**
      * @var string[]
      */
     private $config_param_array = array(
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/DynamicAssurance.php:117]: Add `defaultAssurance`

File: lib/Auth/Process/DynamicAssurance.php [hunk 16]
```diff
@@ -141,67 +152,69 @@ public function process(&$state)
...
+
+        SimpleSAML_Logger::debug(
+            "[DynamicAssurance][process] Assurance mapped after Mappings config: " . var_export($assurance_from_candidates, true)
+        );
 
-        $state['Attributes'][$this->attribute] = array($assurance);
+        // Append the Default Assurance if the Assurance list is empty
+        if(!empty($this->defaultAssurance)
+           && empty($assurance_from_candidates)) {
+            $assurance_from_candidates[] = $this->defaultAssurance;
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/DynamicAssurance.php]: array_merge() ?

**Review by NicolasLiampotis — CHANGES_REQUESTED** [review]

File: lib/Auth/Process/DynamicAssurance.php (hunk 9 above)
Comment (ioigoume) [reply lib/Auth/Process/DynamicAssurance.php]: Done

File: lib/Auth/Process/DynamicAssurance.php (hunk 12 above)
Comment (ioigoume) [reply lib/Auth/Process/DynamicAssurance.php]: Done

File: lib/Auth/Process/DynamicAssurance.php (hunk 13 above)
Comment (ioigoume) [reply lib/Auth/Process/DynamicAssurance.php]: Done

File: lib/Auth/Process/DynamicAssurance.php (hunk 15 above)
Comment (ioigoume) [reply lib/Auth/Process/DynamicAssurance.php:117]: Done

File: lib/Auth/Process/DynamicAssurance.php (hunk 16 above)
Comment (ioigoume) [reply lib/Auth/Process/DynamicAssurance.php]: Done

File: lib/Auth/Process/DynamicAssurance.php (hunk 14 above)
Comment (ioigoume) [reply lib/Auth/Process/DynamicAssurance.php]: Done

File: README.md [hunk 17]
```diff
@@ -13,42 +11,67 @@ The following classes are needed to configure, in order to execute the module.
...
+                'https://refeds.org/assurance/IAP/medium',
+            ),
+            '1.2.840.113612.5.2.2.5' => array(                  // MICS
+                'https://refeds.org/assurance/IAP/low',
+                'https://refeds.org/assurance/IAP/medium',
+            ),
+        ),
+        '/^https:\/\/refeds\.org\/assurance/m',                 // REFEDS passthrough values
+        '/^https:\/\/aarc-community\.org\/assurance/m',         // AARC passthrough values
+    );
```
Comment (NicolasLiampotis) [inline README.md]: 'eduPersonAssurance' => array(
            '1.2.840.113612.5.2.2.1' => array(                  // Classic
                'https://refeds.org/assurance/IAP/low',
                'https://refeds.org/assurance/IAP/medium',
            ),
            '1.2.840.113612.5.2.2.5' => array(                  // MICS
                'https://refeds.org/assurance/IAP/low',
                'https://refeds.org/assurance/IAP/medium',
            ),
            '/^https:\/\/refeds\.org\/assurance/m',                 // REFEDS passthrough values
            '/^https:\/\/aarc-community\.org\/assurance/m',         // AARC passthrough values
        ),

**Review by NicolasLiampotis — CHANGES_REQUESTED** [review]

File: README.md (hunk 11 above)
Comment (ioigoume) [reply README.md:81]: Done

File: README.md (hunk 17 above)
Comment (ioigoume) [reply README.md]: Done

File: README.md (hunk 10 above)
Comment (ioigoume) [reply README.md]: Done

File: README.md (hunk 8 above)
Comment (ioigoume) [reply README.md:8]: Done

File: README.md (hunk 7 above)
Comment (ioigoume) [reply README.md]: Done

File: README.md [hunk 18]
```diff
@@ -1,54 +1,81 @@
...
-  * `attribute`: Optional, a string that defines the name of the attribute that will store the value of LoA. Defaults to `eduPersonAssurance`.
-  * `candidates`: Optional, an array of strings that contains the allowed LoA values.
-  * `defaultAssurance`: Optional, a string to use as the default value of LoA. Defaults to `https://www.example.org/low`.
-  * `defaultElevatedAssurance`: Optional, a string to use as the default value of LoA. Defaults to `https://www.example.org/Substantial`.
-  * `idpPolicies`: Optional, an array of stings that contains the allowed IdP Policies.
-  * `idpTags`: Optional, an array of stings that contains the allowed IdP Tags.
+  * `attribute`: _Optional_, a string that defines the name of the attribute that will store the value of LoA. Defaults to `eduPersonAssurance`.
+  * `attributeMap`: _Optional_,  a map whose `keys` identify attribute names whose values can be mapped to assurance values. Represented LoA attributes, of the authenticating user, contain any of the list of the allowed values, then the module will append these values into the State. Defaults to
+  ```php
+      $assuranceMap = array(
```
Comment (NicolasLiampotis) [inline README.md]: assuranceMap --> attributeMap

File: README.md [hunk 19]
```diff
@@ -1,54 +1,81 @@
...
+                      'vo_test2:IdP Proxy test2' => array(
+                          'https://example.org/LoA#AssuranceLow',
+                      ),
+                  ),
+                  // The Attribute maps to an assurance level. All Assurance values must be underneath the key zero(0)
+                  'voPersonVerifiedEmail' => array(
+                      0 => array(
+                          'https://example.org/LoA#AssuranceLow',
+                      ),
+                  ),
```
Comment (NicolasLiampotis) [inline README.md]: Can't we use a regex instead?
```
'voPersonVerifiedEmail' => array(
    '/.+$/' => array(
        'https://example.org/LoA#AssuranceLow',
    ),
),
```
So the generic attribute map syntax becomes:
```
'<attribute_name> => array(
    '<attribute_value_regex>` => array(
        'assurance_value1',
        'assurance_value2',
    ),
),
```

File: README.md [hunk 20]
```diff
@@ -1,54 +1,81 @@
...
+                'https://refeds.org/assurance/IAP/medium',
+            ),
+        ),
+        '/^https:\/\/refeds\.org\/assurance/m',                 // REFEDS passthrough values
+        '/^https:\/\/aarc-community\.org\/assurance/m',         // AARC passthrough values
+    );
+  ```
+  :warning: The configured Assurance Map array is merged recursively with the default one from above.
+  * `defaultAssurance`: _Optional_, an array containing assurance value(s) to set by default when no assurance information is available. Defaults to `array()`, i.e. no assurance values are added by default.
+  * `idpTagMap`: _Optional_, a map whose `keys` identify IdP Tags which can be mapped to assurance values. IdP Metadata, of the authenticating user, contain any of the list of the allowed values, then the module will append these values into the State. Defaults to `array()`.
```
Comment (NicolasLiampotis) [inline README.md]:
```
_Optional_, a map whose `keys` identify IdP Tags which can be mapped to assurance values. If the metadata of the user's authenticating IdP contain any of the specified tags, then the filter will append these values to the assurance attribute. Defaults to `array()`.
```

File: README.md [hunk 21]
```diff
@@ -1,54 +1,81 @@
...
-  * `attribute`: Optional, a string that defines the name of the attribute that will store the value of LoA. Defaults to `eduPersonAssurance`.
-  * `candidates`: Optional, an array of strings that contains the allowed LoA values.
-  * `defaultAssurance`: Optional, a string to use as the default value of LoA. Defaults to `https://www.example.org/low`.
-  * `defaultElevatedAssurance`: Optional, a string to use as the default value of LoA. Defaults to `https://www.example.org/Substantial`.
-  * `idpPolicies`: Optional, an array of stings that contains the allowed IdP Policies.
-  * `idpTags`: Optional, an array of stings that contains the allowed IdP Tags.
+  * `attribute`: _Optional_, a string that defines the name of the attribute that will store the value of LoA. Defaults to `eduPersonAssurance`.
+  * `attributeMap`: _Optional_,  a map whose `keys` identify attribute names whose values can be mapped to assurance values. Represented LoA attributes, of the authenticating user, contain any of the list of the allowed values, then the module will append these values into the State. Defaults to
```
Comment (NicolasLiampotis) [inline README.md]: * `attributeMap`: _Optional_, a map whose `keys` identify attribute names whose values can be mapped to assurance values. For each identified attribute, you can specify an array of string values to match. For each matching value, the filter appends the specified assurance values to the assurance attribute. Defaults to

File: README.md [hunk 22]
```diff
@@ -1,54 +1,81 @@
...
+                'https://refeds.org/assurance/IAP/low',
+                'https://refeds.org/assurance/IAP/medium',
+            ),
+            '1.2.840.113612.5.2.2.5' => array(                  // MICS
+                'https://refeds.org/assurance/IAP/low',
+                'https://refeds.org/assurance/IAP/medium',
+            ),
+        ),
+        '/^https:\/\/refeds\.org\/assurance/m',                 // REFEDS passthrough values
+        '/^https:\/\/aarc-community\.org\/assurance/m',         // AARC passthrough values
```
Comment (NicolasLiampotis) [inline README.md]: Move lines 26-27 before line 25

**Review by NicolasLiampotis — CHANGES_REQUESTED** [review]

File: README.md (hunk 18 above)
Comment (ioigoume) [reply README.md]: Done

File: README.md (hunk 19 above)
Comment (ioigoume) [reply README.md]: Done

File: README.md (hunk 20 above)
Comment (ioigoume) [reply README.md]: Done

File: README.md (hunk 22 above)
Comment (ioigoume) [reply README.md]: Done

File: README.md (hunk 21 above)
Comment (ioigoume) [reply README.md]: Done

File: README.md [hunk 23]
```diff
@@ -1,54 +1,94 @@
...
+                'https://refeds.org/assurance/IAP/low',
+                'https://refeds.org/assurance/IAP/medium',
             ),
-            'defaultAccurance' => 'https://example.org/LowAssurance',
-            'defaultElevatedAssurance' => 'https://example.org/HighAssurance',
-            'idpPolicies' => array(
-                'example.org:policy01',
-                'example.org:policy02',
+            'pregMatch' => array(
+                '/^https:\/\/refeds\.org\/assurance/m',           // REFEDS passthrough values
```
Comment (NicolasLiampotis) [inline README.md]: Use `#` as regex delimiters to avoid escaping `/`

File: lib/Auth/Process/DynamicAssurance.php [hunk 24]
```diff
@@ -141,67 +172,109 @@ public function process(&$state)
...
 
-        $assurance = $this->defaultAssurance;
+            $preg_match = array();
+            // Check if there is a pregMatch key
+            if(!empty($valAssuranceCandidates['pregMatch'])) {
+                $preg_match = $valAssuranceCandidates['pregMatch'];
+            }
+
+            // Handle any State Attribute having an exact match into configuration
+            foreach ($state['Attributes'][$attribute] as $assurance_oid) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/DynamicAssurance.php]: rename `$assurance_oid` as `$attributeValue` or `$attribute_value` (do we use camel case or underscore case for property names?)

File: lib/Auth/Process/DynamicAssurance.php [hunk 25]
```diff
@@ -141,67 +172,109 @@ public function process(&$state)
...
+                if(is_string($val)) {
+                    $passthrough_values        = preg_grep($valAssuranceCandidates, $state['Attributes'][$attribute]);
+                    if(!empty($passthrough_values)) {
+                        $assurance_from_candidates = array_merge(
+                            $assurance_from_candidates,
+                            $passthrough_values
+                        );
+                    }
+                } elseif(is_array($val)) {  // Regex with list of Assurance values
+                    foreach ($state['Attributes'][$attribute] as $assur_val) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/DynamicAssurance.php]: rename `$assur_val` as `$attributeValues` or `$attribute_values`

File: lib/Auth/Process/DynamicAssurance.php [hunk 26]
```diff
@@ -141,67 +172,109 @@ public function process(&$state)
...
-        if (
-            !empty($state['Attributes']['eduPersonAssurance'])
-            && !empty(array_intersect($state['Attributes']['eduPersonAssurance'], $this->idpPolicies))
-        ) {
-            SimpleSAML_Logger::debug(
-                "[DynamicAssurance] Assurance matches known policy value: "
-                . var_export(array_intersect($state['Attributes']['eduPersonAssurance'], $this->idpPolicies), true)
-            );
-            $assurance = $this->defaultElevatedAssurance;
+        foreach ($this->idpTagMap as $idpTag => $assuranceProfile) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/DynamicAssurance.php]: rename `$assuranceProfile` as `$assuranceValues` or `$assurance_values`

File: README.md [hunk 27]
```diff
@@ -1,54 +1,94 @@
...
-    authproc = array(
-        ...
-        40 => array(
-            'class' => 'assurance:DynamicAssurance',
-            'attribute' => 'eduPersonAssurance',
-            'candidates' => array(
-                'https://refeds.org/profile/sfa',
-                'https://refeds.org/profile/mfa',
+  * `attribute`: _Optional_, a string that defines the name of the attribute that will store the value of LoA. Defaults to `eduPersonAssurance`.
+  * `attributeMap`: _Optional_,  a map whose keys identify attribute names whose values can be mapped to assurance values. For each identified attribute, you can specify an array of string values to match. For each matching value, the filter appends the specified assurance values to the assurance attribute. Defaults to
```
Comment (NicolasLiampotis) [inline README.md]: * `attributeMap`: _Optional_, a map whose keys identify attribute names whose values can be mapped to assurance values. For each identified attribute, you can specify an array of values that will be treated as literal strings for exact matching. It is also possible to use the `pregMatch` key for defining a list of regular expressions that will be matched against the attribute values. For each matching value, the filter appends the specified assurance values to the assurance attribute. Defaults to

**Review by NicolasLiampotis — CHANGES_REQUESTED** [review]

File: lib/Auth/Process/DynamicAssurance.php (hunk 26 above)
Comment (ioigoume) [reply lib/Auth/Process/DynamicAssurance.php]: Done

File: lib/Auth/Process/DynamicAssurance.php (hunk 25 above)
Comment (ioigoume) [reply lib/Auth/Process/DynamicAssurance.php]: Done

File: lib/Auth/Process/DynamicAssurance.php (hunk 24 above)
Comment (ioigoume) [reply lib/Auth/Process/DynamicAssurance.php]: Done

File: README.md (hunk 23 above)
Comment (ioigoume) [reply README.md]: Done

File: README.md (hunk 27 above)
Comment (ioigoume) [reply README.md]: Done

**Review by NicolasLiampotis — APPROVED** [review]
LGTM

**Pull request merged**